### DISCUSSION TITLE
理解 Golang 的函数调用 · /golang-function-call

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#539] [Announcements]
https://draveness.me/golang-function-call

**Comment by wujunze** [comment]
学习了 👍

**Comment by rabbitLove520** [comment]
mark一下

**Comment by lifei6671** [comment]
很有深度。

**Comment by yun-cn** [comment]
Mark

**Comment by yeyuqiu** [comment]
GOLANG CALL STACK BEFORE RETURN 这张图里为什么是 (36)SP ? 应该是 (40)SP 吧

**Comment by draveness** [comment]
> GOLANG CALL STACK BEFORE RETURN 这张图里为什么是 (36)SP ? 应该是 (40)SP 吧

确实写错了..

**Comment by yfz0574** [comment]
“在 C 语言中调用一个函数，函数的参数是通过寄存器和栈传递的”这句话很有问题，这一般发生在X64情况下，X86通常是通过栈传递的，另外，调用约定应该也不是C标准规定的内容，而是由编译器自行实现，所以各家编译器的实现可能会不同。

**Comment by draveness** [comment]
> “在 C 语言中调用一个函数，函数的参数是通过寄存器和栈传递的”这句话很有问题，这一般发生在X64情况下，X86通常是通过栈传递的，另外，调用约定应该也不是C标准规定的内容，而是由编译器自行实现，所以各家编译器的实现可能会不同。

我改成 x86-64 吧，ref: https://en.wikipedia.org/wiki/X86_calling_conventions#List_of_x86_calling_conventions

**Comment by yfz0574** [comment]
“函数是 Go 语言中的一等公民” 这句话用在这里不合适吧？已经有惯用法了，一般是函数式语言用的，在维基百科中有定义。

**Comment by draveness** [comment]
> “函数是 Go 语言中的一等公民” 这句话用在这里不合适吧？已经有惯用法了，一般是函数式语言用的，在维基百科中有定义。

不知道哪里不合适？

> In computer science, a programming language is said to have first-class functions if it treats functions as first-class citizens. This means the language supports passing functions as arguments to other functions, returning them as the values from other functions, and assigning them to variables or storing them in data structures.

+ https://en.wikipedia.org/wiki/First-class_function wikipedia
+ https://blog.golang.org/first-class-functions-in-go-and-new-go golang official blog
+ https://dave.cheney.net/2016/11/13/do-not-fear-first-class-functions  by dave cheney

**Comment by yfz0574** [comment]
> 不知道哪里不合适？

我认错，看来是我误解了。评论有点冒失了。

**Comment by wangxunbang** [comment]
很有收获，给大佬点赞

**Comment by superaitaotao** [comment]
"Go 语言中对于整型和数组类型的参数都是值传递的，也就是在调用函数时会对内容进行拷贝，需要注意的是如果当前数组的大小非常的大"
这里的意思是所有数组成员都会被复制吗？
我尝试了一下，好像就数组地址改了，成员地址并没有改动。
```
func main() {
	a := []int{1,2,3}
	fmt.Printf("before: %p\n", &a)
	fmt.Printf("before: %p\n", &a[0])
	fmt.Printf("before: %p\n", &a[1])
	changeA(a)
	fmt.Printf("after: %p\n", &a)
	fmt.Printf("after: %p\n", &a[0])
	fmt.Printf("after: %p\n", &a[1])
	fmt.Println(a)
}

func changeA(a []int) {
	a[0] = 100
	fmt.Printf("inside: %p\n", &a)
	fmt.Printf("inside: %p\n", &a[0])
	fmt.Printf("inside: %p\n", &a[1])
}
```
请问我哪里理解错了吗？

**Comment by draveness** [comment]
@superaitaotao 

这个并不是数组啊，切片和数组是不一样的，看一下下面的两个例子

+ 数组：https://play.golang.org/p/5AAdqn3kWa-
+ 切片：https://play.golang.org/p/0Dh3rweauMA

**Comment by zwleagle** [comment]
其实都可以理解为值传递，只是指针或引用的时候传递的是地址而已。

**Comment by draveness** [comment]
> @zwleagle
> 其实都可以理解为值传递，只是指针或引用的时候传递的是地址而已。

对，Go 语言就是传值的

**Comment by tidyoux** [comment]
第一个汇编处的注释似乎有几个笔误：
```
my_function:
	pushq	%rbp
	movq	%rsp, %rbp
	movl	%edi, -4(%rbp)    // 取出第一个参数，放到栈上
	movl	%esi, -8(%rbp)    // 取出第二个参数，放到栈上
	movl	-8(%rbp), %eax    // eax = edi = 1 (eax = esi = 2)
	movl	-4(%rbp), %edx    // edx = esi = 2 (edx = edi = 1)
	addl	%edx, %eax        // eax = eax + edx = 1 + 2 = 3 (eax += edx = 3)
	popq	%rbp
```

**Comment by draveness** [comment]
> 第一个汇编处的注释似乎有几个笔误：
> 
> ```
> my_function:
> 	pushq	%rbp
> 	movq	%rsp, %rbp
> 	movl	%edi, -4(%rbp)    // 取出第一个参数，放到栈上
> 	movl	%esi, -8(%rbp)    // 取出第二个参数，放到栈上
> 	movl	-8(%rbp), %eax    // eax = edi = 1 (eax = esi = 2)
> 	movl	-4(%rbp), %edx    // edx = esi = 2 (edx = edi = 1)
> 	addl	%edx, %eax        // eax = eax + edx = 1 + 2 = 3 (eax += edx = 3)
> 	popq	%rbp
> ```

是的，你看的非常细啊

**Comment by seapub** [comment]
是否需要添加以下知识点:  入参从右到左按顺序压栈并拷贝参数。但是参数的计算是从左到右
```go
package main

import "fmt"

func myFunction(a, b int) (int, int) {
	return a + b, a - b
}

func foo(a int) int {
	fmt.Printf("%d\n", a)
	return a
}

func main() {
	myFunction(foo(66), foo(77))
}

// output:
// 66
// 77
```

**Comment by draveness** [comment]
> 是否需要添加以下知识点: 入参从右到左按顺序压栈并拷贝参数。但是参数的计算是从左到右
> 
> ```go
> package main
> 
> import "fmt"
> 
> func myFunction(a, b int) (int, int) {
> 	return a + b, a - b
> }
> 
> func foo(a int) int {
> 	fmt.Printf("%d\n", a)
> 	return a
> }
> 
> func main() {
> 	myFunction(foo(66), foo(77))
> }
> 
> // output:
> // 66
> // 77
> ```

这个事情可以说一下，等我第二次 Review 整体内容时会加进去的

**Comment by fhquthpdw** [comment]
大佬好，首先非常感谢您的系列文章，让我收获很多。有一个疑惑想请教大佬：
```golang
type MyStruct struct {
	i int
}

func myFunction(a MyStruct, b *MyStruct) {
	a.i = 31
	b.i = 41
	fmt.Printf("in my_function - a=(%d, %p) b=(%v, %p)\n", a, &a, b, &b)
}

func main() {
	a := MyStruct{i: 30}
	b := &MyStruct{i: 40}
	fmt.Printf("before calling - a=(%d, %p) b=(%v, %p)\n", a, &a, b, &b)
	myFunction(a, b)
	fmt.Printf("after calling  - a=(%d, %p) b=(%v, %p)\n", a, &a, b, &b)
}
```

在上面这个例子当中，main函数中调用myFunction时，传递的参数有一个b，这个b是一个变量，只不过b的值是一个指针也就是一个内存地址，在myFunction里面接受到这个b的时候，会复制b的值（就是前面提到的指针）生成一个新的变量，理解对吗？所以你在打印输出的时候所写的&b可以理解成是一个指针类型的变量。

如果把上面的例子稍微改写一下：
```golang
type MyStruct struct {
	i int
}

func myFunction(a MyStruct, b *MyStruct) {
	a.i = 31
	b.i = 41
	fmt.Printf("in my_function - a=(%d, %p) b=(%v, %p, %p)\n", a, &a, b, b, &b)
}

func main() {
	a := MyStruct{i: 30}
	b := &MyStruct{i: 40}
	fmt.Printf("before calling - a=(%d, %p) b=(%v, %p, %p)\n", a, &a, b, b, &b)
	myFunction(a, b)
	fmt.Printf("after calling  - a=(%d, %p) b=(%v, %p, %p)\n", a, &a, b, b, &b)
}
```
在打印的地方都把b的值作为指针打印出来了，不难发现都是一样的，所以我可以理解成指针传递的时候，复制的是指针的指针，不知道我的理解对不对。

**Comment by draveness** [comment]
@fhquthpdw 

> 在上面这个例子当中，main函数中调用myFunction时，传递的参数有一个b，这个b是一个变量，只不过b的值是一个指针也就是一个内存地址，在myFunction里面接受到这个b的时候，会复制b的值（就是前面提到的指针）生成一个新的变量，理解对吗？

对

> 在打印的地方都把b的值作为指针打印出来了，不难发现都是一样的，所以我可以理解成指针传递的时候，复制的是指针的指针，不知道我的理解对不对。

没问题

**Comment by KylinMountain** [comment]
请教大神，一开始看GOLANG CALL STACK BEFORE RETURN那张图，没反应过来stack是倒过来的。我重新理解了之后，main是不是先压入myFunction的两个返回值大小进栈，然后压入两个参数入栈，然后调用时候，再压入main的返回地址，myFunction开始后，先弹出栈顶的返回地址存起来，然后按照从左到右的参数大小分别弹出两个参数，然后计算，计算完了，把结果再挨个写入main函数预先压入的两个返回值地址中，然后再读取之前存储的main返回地址，BP再调整到这个返回地址中呢？

**Comment by draveness** [comment]
> 请教大神，一开始看GOLANG CALL STACK BEFORE RETURN那张图，没反应过来stack是倒过来的。我重新理解了之后，main是不是先压入myFunction的两个返回值大小进栈，然后压入两个参数入栈，然后调用时候，再压入main的返回地址，myFunction开始后，先弹出栈顶的返回地址存起来，然后按照从左到右的参数大小分别弹出两个参数，然后计算，计算完了，把结果再挨个写入main函数预先压入的两个返回值地址中，然后再读取之前存储的main返回地址，BP再调整到这个返回地址中呢？

我大概看了一下，感觉你说的没啥问题

**Comment by logikoisto** [comment]
压入的是main函数的返回地址, 这里 我理解 是不是可以说 压入的是 当前调用函数指令的下一条指令的地址 也就是当前pc指针的地址(当取指执行后 pc指针自动加1)也就是说如果例子中 调用完myFunction,函数后还有其他指令 那么压入的其实不是main函数的返回地址

**Comment by draveness** [comment]
> 压入的是main函数的返回地址, 这里 我理解 是不是可以说 压入的是 当前调用函数指令的下一条指令的地址 也就是当前pc指针的地址(当取指执行后 pc指针自动加1)也就是说如果例子中 调用完myFunction,函数后还有其他指令 那么压入的其实不是main函数的返回地址

是的，是执行完 myFunciton 后返回后，回到 main 函数需要执行的指令的地址

**Comment by woofyzhao** [comment]
1. 最后的两个参数与前面的完全不同，调用方 main 函数通过传递这两个参数
====================================================
调用方 main 函数通过[栈]传递这两个参数

2. 参数增加到8个的时候的c源码可以先给出，读起来更顺，看完汇编才知道是:

```
int add(int a1, int a2, ... int a8) {
     return a1+a2+a3+...+a8;
}
```

---

*2020-09-14 UPDATES: 已修复*

**Comment by zhaobu** [comment]
在传递函数参数和接受返回值时是不是需要结合逃逸分析来决策应该使用指针传递还是结构体呢.

**Comment by draveness** [comment]
> 在传递函数参数和接受返回值时是不是需要结合逃逸分析来决策应该使用指针传递还是结构体呢.

逃逸分析指决定值应该分配在栈上还是堆上

**Comment by fengcunwu** [comment]
大佬您好，我想通过函数调用，验证栈空间是由高地址向低地址增长这个问题，但是我发现对一个变量&之后，会发生内存逃逸，如何避免这个问题呢？

**Comment by draveness** [comment]
> 大佬您好，我想通过函数调用，验证栈空间是由高地址向低地址增长这个问题，但是我发现对一个变量&之后，会发生内存逃逸，如何避免这个问题呢？

不知道 `//go:noescape` 对你这个场景有用么

**Comment by fengcunwu** [comment]
> @draveness
> > 大佬您好，我想通过函数调用，验证栈空间是由高地址向低地址增长这个问题，但是我发现对一个变量&之后，会发生内存逃逸，如何避免这个问题呢？
> 
> 不知道 `//go:noescape` 对你这个场景有用么

这个我搜了一下资料，知道是啥意思，但是具体怎么用呢？

**Comment by fengcunwu** [comment]
> @draveness
> > 大佬您好，我想通过函数调用，验证栈空间是由高地址向低地址增长这个问题，但是我发现对一个变量&之后，会发生内存逃逸，如何避免这个问题呢？
> 
> 不知道 `//go:noescape` 对你这个场景有用么

大佬您好，这个对我这个场景没用。

**Comment by MarsonShine** [comment]
关于函数参数与返回至入栈顺序的问题，其实可以从栈的 FILO 出发好理解一些

**Comment by foolenough** [comment]
非常有深度，👍

**Comment by linj-disanbo** [comment]
图 4-2 Go 语言 main 函数的调用栈: BP 应该和32(SP)是同一个位置	

```
0x0018 00024 (main.go:7)	LEAQ	32(SP), BP
```

---

*2021-01-02 UPDATES: 已修复*

**Comment by Imfan** [comment]
> @fengcunwu
> > @draveness
> > > 大佬您好，我想通过函数调用，验证栈空间是由高地址向低地址增长这个问题，但是我发现对一个变量&之后，会发生内存逃逸，如何避免这个问题呢？
> > 
> > 不知道 `//go:noescape` 对你这个场景有用么
> 
> 大佬您好，这个对我这个场景没用。
> 

你要是贴具体的代码可能更好分析

**Comment by AndySpf** [comment]
你好，请问一下函数调用惯例那里，go的调用如果在被调用函数中用上了defer，且返回值未命名，比如：
```
func myFunction() int {
        var a int 
        defer func(){
            a+=1
        }()
	return a
}
```
这种情况下是在return时先把a的值存入栈上对应位置后，再去进行defer里的计算吗？

**Comment by dillanzhou** [comment]
addl	%edx, %eax        // edx = eax + edx = 36
这里注释应该是eax = edx + eax = 36

**Comment by dillanzhou** [comment]
根据 https://github.com/golang/go/issues/40724
golang的下一个版本1.17将修改函数调用的参数传递方式，改为使用寄存器传参。期待大佬届时的更新。

**Comment by dillanzhou** [comment]
看了这篇才知道参数传数组和传切片原来是不一样的，以前一直以为传数组也和切片一样复制了个header进去，感谢大佬。另外“在传递数组或者内存占用非常大的结构体时，我们应该尽量使用指针作为参数类型来避免发生数据拷贝进而影响性能”这段，建议补充一下在数组或结构体不是特别大的时候，应该尽量传值而不是传引用，传引用有很大概率让数据逃逸到堆上从而调用mallocgc，这个对性能影响较大，这是和C语言有较大差异的地方

**Comment by draveness** [comment]
> 你好，请问一下函数调用惯例那里，go的调用如果在被调用函数中用上了defer，且返回值未命名，比如：
> 
> ```
> 
> func myFunction() int {
> 
>         var a int 
> 
>         defer func(){
> 
>             a+=1
> 
>         }()
> 
> 	return a
> 
> }
> 
> ```
> 
> 这种情况下是在return时先把a的值存入栈上对应位置后，再去进行defer里的计算吗？

这个问题，你可以直接编译成汇编看下结果

**Comment by draveness** [comment]
> 根据 [golang/go#40724](https://github.com/golang/go/issues/40724)
> golang的下一个版本1.17将修改函数调用的参数传递方式，改为使用寄存器传参。期待大佬届时的更新。

等 1.17 的正式版出来会做修改的

**Comment by TomatoMr** [comment]
大佬的文章写得太好了。想要学习golang汇编方面的知识，有什么书籍或资料可以推荐吗

**Comment by draveness** [comment]
> @TomatoMr 大佬的文章写得太好了。想要学习golang汇编方面的知识，有什么书籍或资料可以推荐吗

这个我也没系统学过 Go 的汇编，学学 C 语言的调用惯例、能看懂就足够了，深入学习汇编语言的 ROI 其实很低

**Comment by TomatoMr** [comment]
> @draveness
> > @TomatoMr 大佬的文章写得太好了。想要学习golang汇编方面的知识，有什么书籍或资料可以推荐吗
> 
> 这个我也没系统学过 Go 的汇编，学学 C 语言的调用惯例、能看懂就足够了，深入学习汇编语言的 ROI 其实很低

好的，谢谢大佬回复

**Comment by xinnjie** [comment]
可以再讲讲堆栈arg吗

**Comment by draveness** [comment]
> @xinnjie 可以再讲讲堆栈arg吗

没理解你的问题

**Comment by xinnjie** [comment]
> @draveness
> > @xinnjie 可以再讲讲堆栈arg吗
> 
> 没理解你的问题

就是调试时打堆栈的时候，函数的参数列表其实也和这边挺有联系的。比如数组传值，函数参数里的数组会展开

**Comment by lixiang9194** [comment]
请教一个问题，C 语言的方式能够极大地减少函数调用的额外开销
在第一个例子中，C语言main通过寄存器edi、esi传递参数，但是被调函数my_function又把参数从寄存器edi、esi中取出存到栈中，再从栈中存到寄存器eax、edx,最后才使用寄存器相加，并没有避开对栈的操作，反而是多了从寄存器到栈、又从栈到寄存器的操作
go语言直接把参数压栈，然后使用栈相加，我感觉反而go效率更高呢？

**Reply by amchii** [reply]
> 多了从寄存器到栈、又从栈到寄存器的操作

不是一个栈。

**Comment by gongyiling** [comment]
"CPU 访问栈的开销比访问寄存器高几十倍" 这个应该改为：“CPU 访问内存的开销比访问CPU Cache高几十倍”

**Comment by draveness** [comment]
> @gongyiling
> "CPU 访问栈的开销比访问寄存器高几十倍" 这个应该改为：“CPU 访问内存的开销比访问CPU Cache高几十倍”

个人觉得这里问题不大

**Comment by com-wushuang** [comment]
大佬，请教一个问题，在本文开头的例子中BP寄存器的作用是什么？

**Comment by Mstch** [comment]
> @com-wushuang
> 大佬，请教一个问题，在本文开头的例子中BP寄存器的作用是什么？

栈底

**Comment by com-wushuang** [comment]
> @Mstch
> > @com-wushuang
> > 大佬，请教一个问题，在本文开头的例子中BP寄存器的作用是什么？
> 
> 栈底

我想问的就是栈底指针是用来做什么的

**Comment by betafu** [comment]
1、"而函数的返回值是通过 eax 寄存器进行传递的，由于只使用一个寄存器存储返回值，所以 C 语言的函数不能同时返回多个值。"
这里可能不成因果关系（语言规范与编译器实现 是两个方向），我没有找到相关的文献证明这一句话，而且返回值也可以通过其他的寄存器辅助返回（可以尝试下返回具有两个int成员的结构体）。

2、”C 语言同时使用寄存器和栈传递参数，使用 eax 寄存器传递返回值“
这里可能也有问题，对于指针和整数是这样，大于64字节的返回值，可以使用栈返回。

**Comment by yfz0574** [comment]
1.返回值怎么传递,标准中没看到规定,不过在X86的32位平台下常见的确实不大于32位的都是通过eax返回的,64位则会通过EDX:EAX返回,这应该形成了惯例吧.C语言基于某些设计目的,这种细节是不会加以规定的,比如可能违背跨平台保持最高性能这个目的,所以都是各自定义自己的调用约定,然而奇怪的是,调用约定又没有规定返回值怎么传递,而只规定了参数怎么传递,这点疑惑不解,这到底是一种语言厂商或系统厂商的惯例还是CPU厂商的惯例或者规定呢?
2.过大的数据自然无法直接通过寄存器返回了.

**Comment by iam153** [comment]
0x001d 00029 (main.go:8)	MOVQ	$66, (SP)    // 第一个参数
	0x0025 00037 (main.go:8)	MOVQ	$77, 8(SP)   // 第二个参数
这里66和77是存在(SP)到(SP)+16的空间的，但是在myFunction的汇编代码里面（下面两行），看起来66和77是在(SP)+8到(SP)+24的空间。我觉得这里应该是有一个(SP)的变化，但我自己汇编出来的代码里没看到相关语句。想问下博主，这个变化是CALL指令自动做的吗？
	0x0012 00018 (main.go:4)	MOVQ	"".a+8(SP), AX    // AX = 66
	0x0017 00023 (main.go:4)	ADDQ	"".b+16(SP), AX   // AX = AX + 77 = 143

**Comment by iam153** [comment]
> @gongyiling
> "CPU 访问栈的开销比访问寄存器高几十倍" 这个应该改为：“CPU 访问内存的开销比访问CPU Cache高几十倍”

突然有点疑惑，SP、BP指向的数据有没有可能存在CPU的缓存里。换句话说，CPU的缓存里除了寄存器之外还有别的空间吗。

**Comment by sch-git** [comment]
入参的压栈顺序和 C 语言一样，都是从右到左，即第一个参数 66 在栈顶的 SP ~ SP+8，第二个参数存储在 SP+8 ~ SP+16 的空间中。
从描述和图来看 参数和返回值应该都是从左到右入栈才对吧。

**Comment by attenton** [comment]
> @iam153
> > @gongyiling
> > "CPU 访问栈的开销比访问寄存器高几十倍" 这个应该改为：“CPU 访问内存的开销比访问CPU Cache高几十倍”
> 
> 突然有点疑惑，SP、BP指向的数据有没有可能存在CPU的缓存里。换句话说，CPU的缓存里除了寄存器之外还有别的空间吗。

CPU cache和寄存器是两个东西，CPU缓存里不存在寄存器，CPU缓存是和内存差不多的东西，只是离CPU更近，数据传输速率更高，寄存器则是CPU自己内部的存储单元。建议你去看看CSAPP这本书，里面会介绍整个计算机体系结构，其中会介绍CPU命令的执行流程，以及寄存器的实现原理

**Comment by attenton** [comment]
> @sch-git
> 入参的压栈顺序和 C 语言一样，都是从右到左，即第一个参数 66 在栈顶的 SP ~ SP+8，第二个参数存储在 SP+8 ~ SP+16 的空间中。
> 从描述和图来看 参数和返回值应该都是从左到右入栈才对吧。

栈在内存中的地址是从高到低分配的

**Comment by attenton** [comment]
> @lixiang9194
> 请教一个问题，C 语言的方式能够极大地减少函数调用的额外开销
> 在第一个例子中，C语言main通过寄存器edi、esi传递参数，但是被调函数my_function又把参数从寄存器edi、esi中取出存到栈中，再从栈中存到寄存器eax、edx,最后才使用寄存器相加，并没有避开对栈的操作，反而是多了从寄存器到栈、又从栈到寄存器的操作
> go语言直接把参数压栈，然后使用栈相加，我感觉反而go效率更高呢？

“把参数压栈“就是利用寄存器把参数存到栈中，在“使用栈相加的时候“就需要把数据从栈中取回到寄存器中，内存是没有计算功能的

**Comment by iam153** [comment]
> @attenton
> > @iam153
> > > @gongyiling
> > > "CPU 访问栈的开销比访问寄存器高几十倍" 这个应该改为：“CPU 访问内存的开销比访问CPU Cache高几十倍”
> > 
> > 突然有点疑惑，SP、BP指向的数据有没有可能存在CPU的缓存里。换句话说，CPU的缓存里除了寄存器之外还有别的空间吗。
> 
> CPU cache和寄存器是两个东西，CPU缓存里不存在寄存器，CPU缓存是和内存差不多的东西，只是离CPU更近，数据传输速率更高，寄存器则是CPU自己内部的存储单元。建议你去看看CSAPP这本书，里面会介绍整个计算机体系结构，其中会介绍CPU命令的执行流程，以及寄存器的实现原理

抱歉我表达有问题。我的原意是说SP、BP指向的栈数据是否一定是在内存里面，有没有可能在cpu附带的存储空间里（缓存或其他我不知道的部位）。希望您可以回答一下我。
另外，感谢您的建议，最近刚从学校过渡到职场，的确还有很多要学习的。最近还在碎片化了解一些公司的办公系统，简单地了解学习go相关的知识等。过几个月完全熟悉环境后会开始系统地学习一些经典书籍。

**Comment by jieqiyue** [comment]
请问比如说我定义了一个数组很大，它是在堆上分配的。就比如说这样： arr := int [9999]{1,2,3,......999}
，那么请问arr这个变量是分配在哪里的啊？
按照Java语言里面的理解，arr是一个引用，一般分配在栈上，不知道go里面是怎么做的

**Comment by yzbmz5913** [comment]
> @draveness
> > 根据 [golang/go#40724](https://github.com/golang/go/issues/40724)
> > golang的下一个版本1.17将修改函数调用的参数传递方式，改为使用寄存器传参。期待大佬届时的更新。
> 
> 等 1.17 的正式版出来会做修改的

1.17出来啦，坐等大佬更新~

**Comment by xuguoqiang-ascaler** [comment]
那个4-3的图是不是不太对。。。 就是栈上地址的标识

**Comment by codingggggg** [comment]
写的真好，感谢

**Comment by zyj123** [comment]
"结构体在内存中是一片连续的空间，指向结构体的指针也是指向这个结构体的首地址。”
希望写一篇结构体的文章哈。

**Comment by nonbutAworker** [comment]
从上述的汇编代码中我们可以看出，当前函数在执行时首先会将 main 函数中预留的两个返回值地址置成 int 类型的默认值 0，然后根据栈的相对位置获取参数并进行加减操作并将值存回栈中，在 myFunction 函数返回之间，栈中的数据如下图所示


是不是应该是“在 myFunction 函数返回之前”

**Comment by nonbutAworker** [comment]
> @jieqiyue
> 请问比如说我定义了一个数组很大，它是在堆上分配的。就比如说这样： arr := int [9999]{1,2,3,......999}
> ，那么请问arr这个变量是分配在哪里的啊？
> 按照Java语言里面的理解，arr是一个引用，一般分配在栈上，不知道go里面是怎么做的 

较大的对象会被编译器分配到堆上，编译器认为对象较大的条件是大于64kb，一个9999个int类型的数组对象占用的内存空间是8*9999 字节，差不多80kb，所以会被分配到堆上

**Comment by alpha-baby** [comment]
> > @jieqiyue
> > 请问比如说我定义了一个数组很大，它是在堆上分配的。就比如说这样： arr := int [9999]{1,2,3,......999}
> > ，那么请问arr这个变量是分配在哪里的啊？
> > 按照Java语言里面的理解，arr是一个引用，一般分配在栈上，不知道go里面是怎么做的
> 
> 较大的对象会被编译器分配到堆上，编译器认为对象较大的条件是大于64kb，一个9999个int类型的数组对象占用的内存空间是8*9999 字节，差不多80kb，所以会被分配到堆上

大佬可以贴一下：`编译器认为对象较大的条件是大于64kb` 这个代码是在编译器的那个地方判断的呢？

**Comment by nonbutAworker** [comment]
可以查看这个链接，在isSmallMakeSlice函数中判断了对于切片slice来说，什么情况下会触发堆逃逸把对象分配到堆上。
https://stackoverflow.com/questions/42243197/what-is-considered-small-object-in-go-regarding-stack-allocation

AlphaBaby ***@***.***> 于2022年3月5日周六 16:56写道：

**Comment by amchii** [comment]
当前go 1.18编译以下代码

```go
func myFunction(a, b int) (int, int) {
	return a + b, a - b
}

func main() {
	myFunction(66, 77)
}
```

结果已经和文章不一样了：

```assembly
 "".main STEXT size=216 args=0x0 locals=0xa8 funcid=0x0 align=0x0
 ...
 0x0006 00006 (hello.go:7)       SUBQ    $24, SP
 0x000a 00010 (hello.go:7)       MOVQ    BP, 16(SP)
 0x000f 00015 (hello.go:7)       LEAQ    16(SP), BP
 ...
 0x0014 00020 (hello.go:8)       MOVL    $66, AX
 0x0019 00025 (hello.go:8)       MOVL    $77, BX
```

入参也跟C一样使用寄存器，在我的Intel Mac上，会一直使用到R11(AX,BX,CX,DI,SI,R8,R9,R10,R11)共9个寄存器，比C还多。

**Reply by draveness** [reply]
这个版本基于 1.15 的 Go 语言，今年下半年会对整本书的内容做一次修订

**Reply by uniqueni** [reply]
修订的书还在路上吗 大佬

**Comment by aecra** [comment]
C 语言调用栈中多参数传递的汇编源码最后返回值的注释有误。

```C
addl	%edx, %eax        // edx = eax + edx = 36
```

注释中应该是 `eax = eax + edx = 36`。

**Comment by ytlmike** [comment]
想问一下，如果我创建了一个数组变量`[10000]int{1,2.....10000}`，它会分配在堆上，将它作为入参调用一个函数时，会在堆上复制整个数组吗？

**Comment by herrhu97** [comment]
mark

**Comment by Onlyot** [comment]
不太清楚 symbol+offset(SP)这里的SP是伪SP，offset(SP)的SP是硬件SP，在 4.1.1的Go语言中，感觉混用了，是不是统一用一个或者标明一下更清晰呢？

**Comment by lingguo610** [comment]
go语言也是通过寄存器来传递 返回值的

**Reply by lingguo610** [reply]
我看的是 go 1.18.1 版本的

**Reply by draveness** [reply]
新版本改的

**Comment by miaoec** [comment]
mark

**Comment by miaoec** [comment]
```golang
package main

func main() {
	add(3, 4, 5, 6, 7, 8, 9, 10, 61, 7, 8, 9, 10, 10, 10, 10, 10)
}

//go:noinline
func add(a, b, c, d, e, f, g, h, i, j, k, o, p, q, r, s, t int) int {
	return a + b + c + d + e + f + g + h
}
```
``` s
        0x0026 00038 (main.go:4)        JMP     40
        0x0028 00040 (main.go:5)        MOVQ    $7, (SP)
        0x0030 00048 (main.go:5)        MOVQ    $8, 8(SP)
        0x0039 00057 (main.go:5)        MOVQ    $9, 16(SP)
        0x0042 00066 (main.go:5)        MOVQ    $10, 24(SP)
        0x004b 00075 (main.go:5)        MOVQ    $10, 32(SP)
        0x0054 00084 (main.go:5)        MOVQ    $10, 40(SP)
        0x005d 00093 (main.go:5)        MOVQ    $10, 48(SP)
        0x0066 00102 (main.go:5)        MOVQ    $10, 56(SP)
        0x006f 00111 (main.go:5)        MOVL    $3, AX
        0x0074 00116 (main.go:5)        MOVL    $4, BX
        0x0079 00121 (main.go:5)        MOVL    $5, CX
        0x007e 00126 (main.go:5)        MOVL    $6, DI
        0x0083 00131 (main.go:5)        MOVL    $7, SI
        0x0088 00136 (main.go:5)        MOVL    $8, R8
        0x008e 00142 (main.go:5)        MOVL    $9, R9
        0x0094 00148 (main.go:5)        MOVL    $10, R10
        0x009a 00154 (main.go:5)        MOVL    $61, R11
        0x00a0 00160 (main.go:5)        PCDATA  $1, $0
        0x00a0 00160 (main.go:5)        CALL    "".add(SB)
```
似乎现在会用寄存器传参数了，超过9参数个才用栈空间

**Reply by ch0ngsheng** [reply]
go后来改了调用规约，优先用寄存器了

**Comment by dusx1981** [comment]
C 的调用惯例的图，base pointer 在被调用函数栈帧，而返回地址，是在调用函数栈帧，图是不是错了。

**Comment by dusx1981** [comment]
还有，从GO的汇编可以看到，从 myFun 到 main 的返回逻辑，实际上是 Main 来负责的，所以是不是 myFun 的栈帧不应该有返回地址

**Comment by teekoksiang** [comment]
有没有英文版的？

**Comment by JokerDevops** [comment]
怎么理解入栈的顺序是从右到左，我看上面的写的案例。从上到下不是从左边第一个参数，一次往下面传吗。。。

**Reply by JokerDevops** [reply]
看错了，是从右到左。

**Reply by JokerDevops** [reply]
0x001d 00029 (main.go:8)	MOVQ	$66, (SP)    // 第一个参数
	0x0025 00037 (main.go:8)	MOVQ	$77, 8(SP)   // 第二个参数

go 语言这里的编译输出，这不是先把左边的第一个参数入栈了吗。。。

**Reply by JokerDevops** [reply]
C 语言能看的很清楚
main:
	pushq	%rbp
	movq	%rsp, %rbp
	subq	$16, %rsp     // 为参数传递申请 16 字节的栈空间
	movl	$8, 8(%rsp)   // 传递第 8 个参数
	movl	$7, (%rsp)    // 传递第 7 个参数
	movl	$6, %r9d
	movl	$5, %r8d
	movl	$4, %ecx
	movl	$3, %edx
	movl	$2, %esi
	movl	$1, %edi
	call	my_function

**Comment by wanghaolong613** [comment]
学习了 👍

**Comment by Planter-bro** [comment]
买完纸质书，看了三章，对个人而言，可读性相当差，文中截取的源码变量第一次出现时至少要说清楚是什么，我读书是为了提升效率的，这本书并没有帮我节省时间成本，甚至不如直接源码

**Comment by hbyscx001** [comment]
有个疑问是，type HandlerFunc func(*Conn, *Request)这个函数实现了func (f HandlerFunc) ServeHTTP(c *Conn, req *Request)

使用的时候是这样的：http.Handle("/args", http.HandlerFunc(ArgServer))

给人的感觉HandlerFunc具备两个不同的函数实现？
一个是实现是类似new的func(func(*Conn, *Request))
另一个实现是之前定义的func(*Conn, *Request)

**Reply by hbyscx001** [reply]
我怎么感觉改成这种更容易理解：http.Handle("/args", http.HandlerFunc{ArgServer})
或者http.Handle("/args", (http.HandlerFunc)ArgServer)

**Comment by unifreak** [comment]
UPDATE: Go 1.17 开始使用基于寄存器的调用了:

<https://tip.golang.org/doc/go1.17#compiler>:

> Go 1.17 implements a new way of passing function arguments and results using
  registers instead of the stack. Benchmarks for a representative set of Go
  packages and programs show performance improvements of about 5%, and a
  typical reduction in binary size of about 2%. This is currently enabled for
  Linux, macOS, and Windows on the 64-bit x86 architecture (the linux/amd64,
  darwin/amd64, and windows/amd64 ports)."

> Go 1.18 expands the supported platforms to include 64-bit ARM
  (GOARCH=arm64), big- and little-endian 64-bit PowerPC(GOARCH=ppc64, ppc64le),
  as well as 64-bit x86 architecture(GOARCH=amd64) on all operating systems.
